File: lib/services/repositories/service_repository.dart
import '../../models/service_model.dart';
import '../api/api_service.dart';

class ServiceRepository {
  static Future<String?> create(ServiceModel service) async {
    final response = await ApiService.post('/services', service.toJson(), requiresAuth: true);
    if (response['success'] == true && response['service'] != null) {
      return response['service']['id'] as String?;
    }
    return null;
  }

  static Future<ServiceModel?> findById(String id) async {
    final response = await ApiService.get('/services/$id', requiresAuth: true);
    if (response['success'] == true && response['service'] != null) {
      return ServiceModel.fromJson(response['service']);
    }
    return null;
  }

  static Future<List<ServiceModel>> findByClientId(String clientId) async {
    final response = await ApiService.get('/services/client/$clientId', requiresAuth: true);
    if (response['success'] == true && response['services'] != null) {
      final services = response['services'] as List;
      return services.map((s) => ServiceModel.fromJson(s)).toList();
    }
    return [];
  }

  static Future<List<ServiceModel>> findByProfessionalId(String professionalId) async {
    final response = await ApiService.get('/services/professional/$professionalId', requiresAuth: true);
    if (response['success'] == true && response['services'] != null) {
      final services = response['services'] as List;
      return services.map((s) => ServiceModel.fromJson(s)).toList();
    }
    return [];
  }

  static Future<bool> update(String id, Map<String, dynamic> updates) async {
    final response = await ApiService.put('/services/$id', updates, requiresAuth: true);
    return response['success'] == true;
  }

  static Future<bool> delete(String id) async {
    final response = await ApiService.delete('/services/$id', requiresAuth: true);
    return response['success'] == true;
  }
}

File: backend/src/repositories/serviceRepository.js
const Service = require('../models/Service');

class ServiceRepository {
  async create(serviceData) {
    try {
      const service = new Service(serviceData);
      return await service.save();
    } catch (error) {
      throw error;
    }
  }

  async findById(id, populate = true) {
    try {
      let query = Service.findById(id);
      if (populate) {
        query = query.populate('clientId').populate('professionalId');
      }
      return await query;
    } catch (error) {
      throw error;
    }
  }

  async findByClientId(clientId) {
    try {
      return await Service.find({ clientId })
        .populate('professionalId')
        .sort({ createdAt: -1 });
    } catch (error) {
      throw error;
    }
  }

  async findByProfessionalId(professionalId) {
    try {
      return await Service.find({ professionalId })
        .populate('clientId')
        .sort({ createdAt: -1 });
    } catch (error) {
      throw error;
    }
  }

  async update(id, updates) {
    try {
      return await Service.findByIdAndUpdate(
        id,
        { $set: updates },
        { new: true, runValidators: true }
      );
    } catch (error) {
      throw error;
    }
  }

  async delete(id) {
    try {
      return await Service.findByIdAndDelete(id);
    } catch (error) {
      throw error;
    }
  }
}

module.exports = new ServiceRepository();

